text/0004-recoverable-transaction-history.md

```diff
@@ -0,0 +1,351 @@
+
+To fully support recoverable transaction history, clients should use 0x0100 or 0x0101 Authenticated sender memos with all
+outputs that they send to another party, and 0x0200 Destination memos on all change outputs, which are all sent to the change
+subaddress. Change outputs should be sent *even* in the case that the change value is zero, in order that there will be a destination memo.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+A 16-byte *short address hash* is specified:
+
+```
+address_hash = public_address.digest32(“mc-address-hash”)[0..16]
+```
+
+Here `digest32` refers to the merlin-based hash computed using `mc-crypto-digestible` crate.
+
+The *change subaddress* is specified as the subaddress of index 1. It should not
+be used except for change outputs, and this subaddress should not be shared with third parties.
+
+All transactions should have exactly *one* change output, even if the change value is zero.
+
+Three new memo types are specified:
+
+| Memo type bytes | Name                                              |
+| -----------     | -----------                                       |
+| 0x0100          | Authenticated Sender Memo                         |
+| 0x0101          | Authenticated Sender With Payment Request Id Memo |
+| 0x0200          | Destination Memo                                  |
+
+## 0x0100 Authenticated Sender Memo
+
+The 44-byte memo data is laid out as follows:
+
+| Byte range | Item |
+| ---------- | ---- |
+| 0 - 16     | Sender's Address Hash |
+| 16 - 28    | Unused bytes          |
+| 28 - 44    | HMAC                  |
+
+The HMAC is computed using HMAC-SHA512. 
+(HMAC refers to [RFC 2104 HMAC](https://datatracker.ietf.org/doc/html/rfc2104).)
+(SHA512 refers to the 512 bit variant of [SHA-2](https://en.wikipedia.org/wiki/SHA-2).)
+
+The key to this HMAC is the 32-byte shared secret computed from Ristretto key exchange
+using the Sender's spend key, and the Recipient's view key. (For the sender, we multiply
+the senders private spend key corresponding to the subaddress they wish to identify using,
+by the view public key from the recipient's public address. For the recipient, we multiply
+the view private key of the subaddress that recieved the payment by the spend key
+from the public address that the sender used to identify themselves.)
+
+The text to this HMAC is the concatenation of:
+- The domain separation string "mc-memo-mac"
+- The 32-byte public key of the transaction output to which this memo is attached
+- The memo type bytes
+- The first 28 bytes of the memo data (all of the data except the HMAC bytes).
+
+The HMAC output is truncated to 16 bytes.
+
+This HMAC text is computed in the same way for all category 0x01 memos.
+
+(See in code: https://github.com/mobilecoinfoundation/mobilecoin/pull/814/files#diff-681e5a738ac3ba20d8bcb29bcd66424b189075a4a2c1a27b9429b9da78aa7946R14)
```

Comment:
I think MCIPs should follow stack-overflow rules, the answer should be self-contained instead of requiring a link.

(i.e. this should describe how the bytes are hashed, not link to a PR.)

Reply:
I already have a description of that, this is mainly just to help the readers who want to see a reference implementation

Reply:
I can copy paste the hmac algo description from 0x0100 memo, but it's literally exactly the same? I can also just remove the code link